Adjust compiler/rst.vim for sphinx 1.5.1

pavoljuhas opened the pull request:
It seems that compiler/rst.vim has no active maintainer 
https://groups.google.com/d/msg/vim_dev/BcDThxEkwNA/n0Fqx9COAgAJ
so I am posting a proposed change here.  The new errorformat appears to work 
with output from sphinx 1.5.1 that is produced for one of my Python packages - 
see example in [sphinx-errors.txt](https://github.com/vim/vim/files/886026/sphinx-errors.txt).

I am however not sure if compiler/rst.vim is intended for sphinx or if it should also
work with some other rst documentation tool.

pavoljuhas committed:
Adjust compiler/rst.vim for sphinx 1.5.1.

Also add errorformat fields to capture errors in Python source
docstrings.  These match only the source file and display
further details in non-matching search-text field '%s'.

pavoljuhas committed:
Use :CompilerSet command in compiler plugin.

brammool commented:
Pavol Juhas wrote:

> It seems that compiler/rst.vim has no active maintainer 
> https://groups.google.com/d/msg/vim_dev/BcDThxEkwNA/n0Fqx9COAgAJ
> so I am posting a proposed change here.  The new errorformat appears to work 
> with output from sphinx 1.5.1 that is produced for one of my Python packages - 
> see example in [sphinx-errors.txt](https://github.com/vim/vim/files/886026/sphinx-errors.txt).
> 
> I am however not sure if compiler/rst.vim is intended for sphinx or if it should also
> work with some other rst documentation tool. 

That looks fine.  How about adding a comment about which lines are for
which compiler?

-- 
"Marriage is a wonderful institution...
but who wants to live in an institution?"
 - Groucho Marx

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

pavoljuhas commented:
> That looks fine.  How about adding a comment about which lines are for
which compiler?

I have updated the whole `errorformat` to match the current sphinx output.  I am not aware of other tools for compiling rst files.  Before version 1.0.8 sphinx produced output corresponding to the current errorformat in compiler/rst.vim:
```
/tmp/s1/index.rst:11: (ERROR/3) Undefined substitution referenced: "Grxxx".
```

This PR adjusts for error output from later sphinx versions which is
```
/tmp/s1/index.rst:11: WARNING: Undefined substitution referenced: "Grxxx".
```

I have also added 2 new errorformat patterns for errors produced when sphinx looks at docstrings in Python sources.  These do not show line numbers, but rather offsets from a start of some docstrings
```
/tmp/diffpy.structure/src/diffpy/structure/atom.py:docstring of diffpy.structure.atom.Atom:13: ERROR: Unexpected indentation.
```
Such lines are captured with the `\%f\\:%s:\ %tRROR:\ %m` pattern.  That pattern  indentifies only the source file because the `%s` pattern does not match.  However, `:clist` and `:cwindow` are then still useful for finding the offending source line:

```
/tmp/diffpy.structure/src/diffpy/structure/symmetryutilities.py|^\Vdocstring of diffpy.structure.symmetryutilities.ExpandAsymmetricUnit:6\$| Unexpected indentation.
```


Should I just add comment that rst.vim is for sphinx 1.0.8 or later?

brammool commented:
Pavol Juhas wrote:

> > That looks fine.  How about adding a comment about which lines are for which compiler?
> 
> I have updated the whole `errorformat` to match the current sphinx
> output.  I am not aware of other tools for compiling rst files.
> Before version 1.0.8 sphinx produced output corresponding to the
> current errorformat in compiler/rst.vim:
> ```
> /tmp/s1/index.rst:11: (ERROR/3) Undefined substitution referenced: "Grxxx".
> ```
> 
> This PR adjusts for error output from later sphinx versions which is
> ```
> /tmp/s1/index.rst:11: WARNING: Undefined substitution referenced: "Grxxx".
> ```
> 
> I have also added 2 new errorformat patterns for errors produced when
> sphinx looks at docstrings in Python sources.  These do not show line
> numbers, but rather offsets from a start of some docstrings
> ```
> /tmp/diffpy.structure/src/diffpy/structure/atom.py:docstring of diffpy.structure.atom.Atom:13: ERROR: Unexpected indentation.
> ```
> Such lines are captured with the `\%f\\:%s:\ %tRROR:\ %m` pattern.
> That pattern  indentifies only the source file because the `%s`
> pattern does not match.  However, `:clist` and `:cwindow` are then
> still useful for finding the offending source line:
> 
> ```
> /tmp/diffpy.structure/src/diffpy/structure/symmetryutilities.py|^\Vdocstring of diffpy.structure.symmetryutilities.ExpandAsymmetricUnit:6\$| Unexpected indentation.
> ```
> 
> 
> Should I just add comment that rst.vim is for sphinx 1.0.8 or later?

That would help.  It's good for users to know what compiler the error
messages are coming from.

-- 
Hear about the guy who played a blank tape at full blast?
The mime next door went nuts.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

pavoljuhas committed:
Add comment on compatible sphinx versions.

pavoljuhas commented:
> That would help.  It's good for users to know what compiler the error
messages are coming from.

OK - here is my take on that 1da8294de1d073e70d8ede22d798e3da2de4cce6.

brammool commented:
Thanks, I'll include it.
